File: VB/MvcDashboardUseDifferentCaches/App_Start/DashboardConfig.vb
Imports System
Imports System.Web.Hosting
Imports System.Web.Routing
Imports DevExpress.DashboardCommon
Imports DevExpress.DashboardWeb
Imports DevExpress.DashboardWeb.Mvc
Imports DevExpress.DataAccess.Excel
Imports DevExpress.DataAccess.Sql

Namespace MvcDashboardUseDifferentCaches
	Public Module DashboardConfig
		Public Sub RegisterService(ByVal routes As RouteCollection)
			routes.MapDashboardRoute("dashboardControl", "DefaultDashboard")

			Dim dashboardFileStorage As New DashboardFileStorage("~/App_Data/Dashboards")
			DashboardConfigurator.Default.SetDashboardStorage(dashboardFileStorage)

			' Uncomment this string to allow end users to create new data sources based on predefined connection strings.
			'DashboardConfigurator.Default.SetConnectionStringsProvider(new DevExpress.DataAccess.Web.ConfigFileConnectionStringsProvider());

			Dim dataSourceStorage As New DataSourceInMemoryStorage()

			' Registers an SQL data source.
			Dim sqlDataSource As New DashboardSqlDataSource("SQL Data Source", "NWindConnectionString")
			Dim query As SelectQuery = SelectQueryFluentBuilder.AddTable("SalesPerson").SelectAllColumns().Build("Sales Person")
			sqlDataSource.Queries.Add(query)
			dataSourceStorage.RegisterDataSource("sqlDataSource", sqlDataSource.SaveToXml())

			DashboardConfigurator.Default.SetDataSourceStorage(dataSourceStorage)

			AddHandler DashboardConfigurator.Default.CustomParameters, AddressOf Default_CustomParameters
		End Sub

		Private Sub Default_CustomParameters(ByVal sender As Object, ByVal e As CustomParametersWebEventArgs)
			e.Parameters.Add(New DashboardParameter("UniqueCacheParam", GetType(Guid), CacheManager.UniqueCacheParam))
		End Sub
	End Module
End Namespace

File: CS/MvcDashboardUseDifferentCaches/App_Start/DashboardConfig.cs
using System;
using System.Web.Routing;
using DevExpress.DashboardCommon;
using DevExpress.DashboardWeb;
using DevExpress.DashboardWeb.Mvc;
using DevExpress.DataAccess.Sql;

namespace MvcDashboardUseDifferentCaches
{
    public static class DashboardConfig {
        public static void RegisterService(RouteCollection routes) {
            routes.MapDashboardRoute("dashboardControl", "DefaultDashboard");

            DashboardFileStorage dashboardFileStorage = new DashboardFileStorage("~/App_Data/Dashboards");
            DashboardConfigurator.Default.SetDashboardStorage(dashboardFileStorage);

            // Uncomment this string to allow end users to create new data sources based on predefined connection strings.
            //DashboardConfigurator.Default.SetConnectionStringsProvider(new DevExpress.DataAccess.Web.ConfigFileConnectionStringsProvider());
            
            DataSourceInMemoryStorage dataSourceStorage = new DataSourceInMemoryStorage();
            
            // Registers an SQL data source.
            DashboardSqlDataSource sqlDataSource = new DashboardSqlDataSource("SQL Data Source", "NWindConnectionString");
            SelectQuery query = SelectQueryFluentBuilder
                .AddTable("SalesPerson")
                .SelectAllColumns()
                .Build("Sales Person");
            sqlDataSource.Queries.Add(query);
            dataSourceStorage.RegisterDataSource("sqlDataSource", sqlDataSource.SaveToXml());

            DashboardConfigurator.Default.SetDataSourceStorage(dataSourceStorage);

            DashboardConfigurator.Default.DataSourceCacheKeyCreated += Default_DataSourceCacheKeyCreated;               
        }

        private static void Default_DataSourceCacheKeyCreated(object sender, DataSourceCacheKeyCreatedEventArgs e) {
            e.Key.CustomData.Add("SessionId", CacheManager.UniqueCacheParam.ToString());
        }
    }
}